### PR TITLE
[1249] Register a mentor who has been registered before

### PR DESCRIPTION
**Pull request by mooktakim**
### Context

Ticket: https://github.com/DFE-Digital/register-ects-project-board/issues/1249

### Changes proposed in this pull request

* Register mentor wizard extended to allow registering a mentor who has been registered before.

### CODE REVIEW

File: app/views/schools/register_mentor_wizard/started_on.html.erb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    "When did or when will #{@mentor.full_name} start mentoring ECTs at your school?",
+  error: @wizard.current_step.errors.present?,
+  backlink_href: @wizard.previous_step_path,
+) %>
+
+<%= form_with(model: @wizard.current_step, url: @wizard.current_step_path) do |f| %>
+  <%= content_for(:error_summary) { f.govuk_error_summary } %>
+
+  <%= f.govuk_date_field :started_on,
+                     width: "two-thirds",
```
Comment (peteryates) [inline app/views/schools/register_mentor_wizard/started_on.html.erb:12]: I don't think `width` will do anything here as dates are already constrained.

File: spec/features/schools/mentors/register_mentor/edge_cases/school_led_ect_spec.rb [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+  end
+
+  def and_mentor_has_mentorship_with_new_school
+    expect(@existing_mentor_at_school_period.reload.finished_on).not_to be_nil
+    expect(@teacher.mentor_at_school_periods.count).to eq(2)
+
+    new_mentor_at_school_period = @teacher.mentor_at_school_periods.excluding(@existing_mentor_at_school_period).last
+    expect(new_mentor_at_school_period.started_on).to eq(@mentor_start_date)
+    expect(new_mentor_at_school_period.finished_on).to be_nil
+    expect(new_mentor_at_school_period.training_periods.count).to eq(0)
```
Comment (peteryates) [inline spec/features/schools/mentors/register_mentor/edge_cases/school_led_ect_spec.rb:187]: I think this is going to break once #982 is merged.

**Review by peteryates — APPROVED** [review]
This looks great, left some really minor comments. Tests are very thorough 👌

File: app/services/teachers/mentor_funding_eligibility.rb [hunk 3]
```diff
@@ -6,10 +6,17 @@ def initialize(trn:)
       @teacher = Teacher.find_by(trn:)
     end
 
+    # TODO: `mentor_became_ineligible_for_funding_on` is a date,
+    # should we pass in a date to determine eligibility? (Date.today or "mentoring start date")
```
[the remark this answers is not in the document]
Comment (mooktakim) [reply app/services/teachers/mentor_funding_eligibility.rb:10]: this is where it gets a bit confusing. what if we have mentor start date in the past, should we not check if mentor was inelgible on that date?

File: app/wizards/schools/register_mentor_wizard/mentor.rb [hunk 4]
```diff
@@ -71,11 +73,70 @@ def ect
...
+        ::Teachers::MentorFundingEligibility.new(trn:).ineligible?
+      end
+
+      def ect_lead_provider
+        ect_training_service.latest_lead_provider if ect
+      end
+
+      def ect_eoi_lead_provider
+        ect_training_service.latest_eoi_lead_provider
       end
```
Comment (ltello) [inline app/wizards/schools/register_mentor_wizard/mentor.rb:128]: Please, can we declare the methods in alphabetical order? Otherwise, it is difficult to read.

**Review by ltello — CHANGES_REQUESTED** [review]

File: app/wizards/schools/register_mentor_wizard/wizard.rb [hunk 5]
```diff
@@ -66,6 +71,12 @@ def allowed_steps
             return steps unless mentor.email
             return steps + %i[change_email_address cant_use_changed_email cant_use_email] if mentor.cant_use_email?
 
+            steps << :mentoring_at_new_school_only
+            steps << :started_on
+            steps << :previous_training_period_details
+            steps << :programme_choices
+            steps << :lead_provider
```
Comment (ltello) [inline app/wizards/schools/register_mentor_wizard/wizard.rb:78]: I think this is incomplete/or breaking the purpose of this method to return the allowed steps a user can go from a given step.

File: app/wizards/schools/register_mentor_wizard/wizard.rb [hunk 6]
```diff
@@ -23,6 +23,11 @@ class Wizard < ApplicationWizard
             review_mentor_details: ReviewMentorDetailsStep,
             review_mentor_eligibility: ReviewMentorEligibilityStep,
             trn_not_found: TRNNotFoundStep,
+            mentoring_at_new_school_only: MentoringAtNewSchoolOnlyStep,
+            started_on: StartedOnStep,
+            previous_training_period_details: PreviousTrainingPeriodDetailsStep,
+            programme_choices: ProgrammeChoicesStep,
+            lead_provider: LeadProviderStep,
```
Comment (ltello) [inline app/wizards/schools/register_mentor_wizard/wizard.rb:30]: all steps are kept in alphabetical order for quickly search and find it. Please, can we keep that order?

File: spec/features/schools/mentors/register_mentor/edge_cases/not_mentoring_at_new_school_only_spec.rb [hunk 7]
```diff
@@ -0,0 +1,180 @@
+RSpec.describe 'Registering a mentor', :js do
+  include_context 'test trs api client'
+
+  let(:trn) { '3002586' }
```
[the remark this answers is not in the document]
Comment (mooktakim) [reply spec/features/schools/mentors/register_mentor/edge_cases/not_mentoring_at_new_school_only_spec.rb:4]: mainly to keep consistent with other tests we have as there's a lot of edge cases

File: app/views/schools/register_mentor_wizard/check_answers.html.erb [hunk 8]
```diff
@@ -1,40 +1,69 @@
...
-        row.with_key(text: 'Lead provider')
-        row.with_value(text: ECTAtSchoolPeriods::Training.new(@wizard.ect).latest_lead_provider_name )
-      end
+  if @mentor.lead_provider
+    summary_list.with_row do |row|
+      row.with_key(text: "Lead provider")
+      row.with_value(text: @mentor.lead_provider.name)
+      row.with_action(
+        text: "Change",
+        href: (schools_register_mentor_wizard_lead_provider_path),
```
[the remark this answers is not in the document]
Comment (avinhurry) [reply app/views/schools/register_mentor_wizard/check_answers.html.erb:61]: Yes it can get very messy, I was wondering if its one that we wanted to card up and address at some point or if we were happy with leaving it as is.

**Review by ltello — CHANGES_REQUESTED** [review]

**Comment by peteryates** [comment]
This PR is more than a fortnight old and quite big, maybe we can address the important feedback now and add issues to come back and finish off the smaller/aesthetic things?

File: app/wizards/schools/register_mentor_wizard/started_on_step.rb [hunk 9]
```diff
@@ -0,0 +1,55 @@
+module Schools
+  module RegisterMentorWizard
+    class StartedOnStep < Step
+      attribute :started_on, :date
```
[the remark this answers is not in the document]
Comment (mooktakim) [reply app/wizards/schools/register_mentor_wizard/started_on_step.rb:4]: this serializes correctly as thats what the `attribute` type meant to do.

**Review by avinhurry — APPROVED** [review]
Looks good to me 👍🏽

**Review by ltello — APPROVED** [review]

File: app/wizards/schools/register_mentor_wizard/started_on_step.rb [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+      end
+
+      def started_on_cannot_be_before_previous_started_and_finished_dates
+        return if errors.any?
+
+        date = mentor.previous_school_mentor_at_school_periods.pluck(:started_on, :finished_on).flatten.compact.max
+        return unless date
+
+        if started_on.before?(date.next_day)
+          errors.add(:started_on, "#{mentor.full_name} was registered as a mentor #{date.to_fs(:govuk)}. Enter a later date.")
```
Comment (claire-hughez) [inline app/wizards/schools/register_mentor_wizard/started_on_step.rb:50]:
```suggestion
          errors.add(:started_on, "#{mentor.full_name} was registered as a mentor at their last school starting on the #{date.to_fs(:govuk)}. Enter a later date.")
```

File: app/views/schools/register_mentor_wizard/started_on.html.erb [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+  error: @wizard.current_step.errors.present?,
+  backlink_href: @wizard.previous_step_path,
+) %>
+
+<%= form_with(model: @wizard.current_step, url: @wizard.current_step_path) do |f| %>
+  <%= content_for(:error_summary) { f.govuk_error_summary } %>
+
+  <%= f.govuk_date_field :started_on,
+                     legend: {
+                       text: "Start date",
```
[the remark this answers is not in the document]
Comment (mooktakim) [reply app/views/schools/register_mentor_wizard/started_on.html.erb:13]: fix
